lib/Transforms/Instrumentation/GeneratePartialFnPass.cc

```diff
@@ -0,0 +1,162 @@
+
+    bool changed = false;
+    for (auto &F : M) {
+      if (!F.empty()) {
+        changed |= runOnFunction(SBI, F);
+      }
+    }
+    return changed;
+  }
+
+  bool runOnFunction(SeaBuiltinsInfo &SBI, Function &F) {
+    // Checks that this is a partial function, with a Boolean return type.
+    if (!isPartialFn(F))
+      return false;
+
+    if (!F.getReturnType()->isIntegerTy(1)) {
+      LOG("partial-fn", errs() << "Partial function without Bool return type: "
+                               << F.getName().str() << "\n";);
+      return false;
+    }
+
+    // Provides a placeholder implementation to each partial function stub.
+    bool changed = false;
+    for (auto &BB : F) {
+      IRBuilder<> pfnBuilder(&BB);
+
+      for (auto &I : BB) {
+        // Does I call a user defined function.
+        CallBase *CB = dyn_cast<CallBase>(&I);
+        if (!CB || SBI.isSeaBuiltin(*CB))
+          continue;
+
+        // Is the callee defined.
+        Function *CF = CB->getCalledFunction();
+        if (!CF || !CF->isDeclaration())
+          continue;
+
+        // Does the function return an integer?
+        if (!CF->getReturnType() || !CF->getReturnType()->isIntegerTy())
+          continue;
+
+        // Records changes.
+        changed = true;
+        LOG("partial-fn", errs() << "Found external call from partial function "
+                                 << F.getName().str() << " to "
+                                 << CF->getName().str() << ".\n");
+
+        // Renames the stub so that it can be identified in other passes.
+        CF->setName(std::string(PARTIAL_FN_STUB_PREFIX) + CF->getName().str());
+
+        // The stub should have a body. The body will force the ModuleHornify
+        // pass to generate a call to the stub. Consequently, arguments passed
+        // to the stub will be live within the partial function that calls it.
+        BasicBlock *block = BasicBlock::Create(CF->getContext(), "entry", CF);
+        IRBuilder<> stubBuilder(block);
+        stubBuilder.CreateRet(ConstantInt::get(CF->getReturnType(), 0));
+
+        // Ensures that the stub is not optimized away by sequential LLVM
+        // passes.
+        CF->addFnAttr(Attribute::NoInline);
+        CF->addFnAttr(Attribute::OptimizeNone);
+
+        // Within the partial function, assume that the call to the stub never
+        // happens. This allows for the body to be synthesized, starting from
+        // the location of the call.
+        pfnBuilder.SetInsertPoint(&I);
+        pfnBuilder.CreateCall(m_assumeFn, pfnBuilder.getFalse());
+      }
+    }
+    return changed;
+  }
+
+  virtual StringRef getPassName() const { return "GeneratePartialFn"; }
+};
+
+char GeneratePartialFnPass::ID = 0;
+
+} // namespace
+
+llvm::Pass *createGeneratePartialFnPass() {
+  return new GeneratePartialFnPass();
+}
+
+} // namespace seahorn
+
+using namespace seahorn;
+INITIALIZE_PASS(GeneratePartialFnPass, "partial-fn-generator",
```

Comment:
short pass name should match string used in `LOG`. better use `DOG` macro

Reply:
What is the `DOG` macro?

Reply:
Debug LOG. grep code for its definition